metriq_gym/benchmarks/mirror_circuits.py

```diff
@@ -0,0 +1,386 @@
+
+    return qc
+
+
+def generate_mirror_circuit(
+    num_layers: int,
+    two_qubit_gate_prob: float,
+    connectivity_graph: rx.PyGraph,
+    two_qubit_gate_name: str = "CNOT",
+    seed: int | None = None,
+) -> tuple[QuantumCircuit, str]:
+    """Generate a mirror circuit with specified parameters.
+
+    Creates a quantum circuit with a mirror structure: forward layers of random
+    Clifford operations, a middle layer of random Paulis, then the inverse of
+    the forward layers. The expected outcome is computed by ideal simulation,
+    ensuring accurate benchmarking against the theoretical prediction.
+
+    Args:
+        num_layers: The number of random Clifford layers to be generated.
+        two_qubit_gate_prob: Probability of a two-qubit gate being applied (0.0 to 1.0).
+        connectivity_graph: The connectivity graph of the backend.
+        two_qubit_gate_name: Name of two-qubit gate to use ("CNOT" or "CZ").
+        seed: Seed for generating randomized mirror circuit.
+
+    Returns:
+        A tuple of (mirror_circuit, expected_bitstring) where expected_bitstring
+        is computed from noiseless simulation of the circuit.
+
+    Raises:
+        ValueError: If parameters are invalid.
+    """
+    if not 0 <= two_qubit_gate_prob <= 1:
+        raise ValueError("two_qubit_gate_prob must be between 0 and 1")
+
+    try:
+        TwoQubitGateType(two_qubit_gate_name)
+    except ValueError:
+        raise ValueError(f"two_qubit_gate_name must be one of {list(TwoQubitGateType)}")
+
+    random_state = random.RandomState(seed)
+    num_qubits = len(connectivity_graph.node_indices())
+
+    if num_qubits == 0:
+        qc = QuantumCircuit(1)
+        qc.measure_all()
+        return qc, "0"
+
+    qc = QuantumCircuit(num_qubits)
+
+    initial_clifford_layer = random_single_cliffords(connectivity_graph, random_state, seed)
+    qc.compose(initial_clifford_layer, inplace=True)
+
+    forward_layers = []
+    for layer_idx in range(num_layers):
+        pauli_layer = random_paulis(connectivity_graph, random_state)
+        qc.compose(pauli_layer, inplace=True)
+        forward_layers.append(pauli_layer)
+
+        selected_edges = edge_grab(two_qubit_gate_prob, connectivity_graph, random_state)
+        layer_seed = seed + layer_idx if seed is not None else None
+        clifford_layer = random_cliffords(
+            selected_edges, random_state, two_qubit_gate_name, layer_seed
+        )
+        qc.compose(clifford_layer, inplace=True)
+        forward_layers.append(clifford_layer)
+
+    middle_pauli = random_paulis(connectivity_graph, random_state)
+    qc.compose(middle_pauli, inplace=True)
+
+    for layer in reversed(forward_layers):
+        qc.compose(layer.inverse(), inplace=True)
+
+    qc.compose(initial_clifford_layer.inverse(), inplace=True)
+
+    qc.measure_all()
+
+    sim_circuit = qc.copy()
+    sim_circuit.remove_final_measurements()
+
+    try:
+        statevector = Statevector(sim_circuit)
```

Comment:
Is a simulation actually needed here? Or in a noiseless scenario, we can assume the expected bitstring is always 0..0?